### PR TITLE
Test fetch v2

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR rewrites the test fetcher util to be more accurate in the tests collection, and also comes with a restriction on the tests run when a large amount of tests are picked when modifying a core file (like modeling_utils).

The code that extracts the dependencies of a given module now inspects the inits to pinpoint the exact location of imported objects. So for instance if a test file has an import `from transformers import BertModel`, this new version will detect a dependency on `transformers/models/bert/modeling_bert.py`. As a comparison, the previous version stopped at `transformers/__init__.py`. This removes the need for all the complex logic that tried to match a given file with its corresponding tests, we now just look at the dependencies of the test file.

The second change is that when a given file is seen to trigger too many model tests (current trigger is set at half the models, it can evolve), it will only keep the tests relative to a given list of important models. If a PR changes many modeling files, all the tests for those models will still run, but if a PR only changes modeling_utils (for instance), this will trigger the core model tests only. The list of important models is built using:
- the most downloaded models in the last 30 days
- making sure each pipeline has a model in that list

To bypass this rule, one can add a special command in a commit message (circleCI does not have access to labels, so I can't rely on that):
- Including [skip ci] or [ci skip] or [circleci skip] or [skip circleci] or any variants with - or _ instead of a space will skip all tests
- Including [test all models] or any variant with the words in another order and/or with - or _ instead of a space will run all tests found without filtering on important models.
- Including [test all] or [all test] or any variants with - or _ instead of a space will run all tests.

A couple of adjustments to Transformers should be done (in follow-up PRs) to have the test fetcher be more accurate and more efficient:
- make sure all inits don't define any objects. Most of our inits only import all the stuff, and the test fetcher assumes they are all like that. Some inits (like `pipeline/__init__.py`) define real objects, it would be best to move them to a submodule.
- make sure test files test one thing: for instance `test_modeling_common.py` contains both the common tests and the test of the modeling_utils module. It would be best to split those in two files.

Lastly, this PR adds lots of tests to make sure future work doesn't break the test fetcher :-)

To see how the test fetcher behaves on some examples:
- for a modification in modeling_opt.py: only test_modeling_opt is run  [[fetch summary](https://app.circleci.com/pipelines/github/huggingface/transformers/60906/workflows/79cfaf18-d0da-4a5d-8bc6-fd8599b63468/jobs/747209/artifacts)] [[job page](https://app.circleci.com/pipelines/github/huggingface/transformers/60710/workflows/f43d1235-3337-484f-b79a-9260e24e3664)]
- for a modification in modeling_bert.py (which is imported in all the tests basically) all tests using BERT are run, but filtered to the list of important models [[fetch summary](https://app.circleci.com/pipelines/github/huggingface/transformers/60906/workflows/79cfaf18-d0da-4a5d-8bc6-fd8599b63468/jobs/747209/artifacts)] [[job page](https://app.circleci.com/pipelines/github/huggingface/transformers/60906/workflows/f85bd255-a01a-49af-a852-b7a00e13aad3)]
- for a modification in a pipeline file: all model tests are run, filtered to the list of important models [[fetch summary](https://app.circleci.com/pipelines/github/huggingface/transformers/60915/workflows/ab1dc699-5ef7-4b33-bcab-62f76688e9f4/jobs/747362/artifacts)] [[job page](https://app.circleci.com/pipelines/github/huggingface/transformers/60910/workflows/3f9df353-e3b0-490c-9724-f6ef59df5599)]
- for a modification in the main `__init__.py` all tests are run, but filtered to the list of important models [[fetch summary](https://app.circleci.com/pipelines/github/huggingface/transformers/60910/workflows/1f8039a1-af57-477d-bd20-43324d574549/jobs/747284/artifacts)] [[job page](https://app.circleci.com/pipelines/github/huggingface/transformers/60915/workflows/9615653e-b1e5-4c74-9668-ca7983bc3b68)]
- for a modification in the `setup.py` all tests are run [[fetch summary](https://app.circleci.com/pipelines/github/huggingface/transformers/60913/workflows/fd3a907c-98a3-40b0-8423-d1346b23498c/jobs/747326/artifacts)] [[job page](https://app.circleci.com/pipelines/github/huggingface/transformers/60913/workflows/809e8564-7df0-4321-8d7c-4e68c1d123a7)]

### CODE REVIEW

**Comment by ydshieh** [comment]
I guess it's ready to be reviewed ..?

**Comment by sgugger** [comment]
No I haven't finished this PR yet.

File: utils/tests_fetcher.py [hunk 1]
```diff
@@ -13,6 +13,27 @@
...
+all modifications in a PR, not just the last commit) but excludes modifications that are on docstrings or comments
+only.
+
+Stage 2: Extract the tests to run. This is done by looking at the imports in each module and test file: if module A
+imports module B, then changing module B impacts module A, so the tests using module A should be run. We thus get the
+dependencies of each model and the recursively builds the 'reverse' map of dependencies to get all modules and tests
+impacted by a given file. We then only keep the tests (and only the code models tests if there are too many modules).
+
+Caveats:
+  - This module only filter tests by files (not individual tests) so it's better to have tests for different things
```
Comment (LysandreJik) [inline utils/tests_fetcher.py:31]:
```suggestion
  - This module only filters tests by files (not individual tests) so it's better to have tests for different things
```

File: utils/tests_fetcher.py [hunk 2]
```diff
@@ -79,17 +123,21 @@ def get_all_tests():
...
-    model_tests_folders = os.listdir(os.path.join(test_root_dir, "models"))
-    model_test_folders = sorted(filter(os.path.isdir, [f"tests/models/{x}" for x in model_tests_folders]))
+    model_test_folders = os.listdir(PATH_TO_TESTS / "models")
+    model_test_folders = [f"tests/models/{f}" for f in model_test_folders if "__pycache__" not in f]
+    model_test_folders = sorted([f for f in model_test_folders if (PATH_TO_REPO / f).is_dir()])
 
     tests.remove("tests/models")
+    # Sagemaker tests are not meant to be run on the CI.
+    if "tests/sagemaker" in tests:
+        tests.remove("tests/sagemaker")
```
Comment (LysandreJik) [inline utils/tests_fetcher.py:140]: to verify with @philschmid

**Review by LysandreJik — APPROVED** [review]
Cool, LGTM! Looking forward to seeing it run!

**Comment by ydshieh** [comment]
Hi @sgugger . Thank you a lot for working on this important task! I feel it's better for me to look this work in depth, and I tried to play with the test fetcher (on `main` and on this PR) to understand it better. 

However, the first thing I tried (by following some sentences you mentioned) makes me a somehow confused. Here is what I saw:

- On the two branches `main` (or a new branch from it) and `test_fetch_v2`, do the following steps:
  - change the test file `tests/models/bert/test_modeling_bert.py`  (simply adding some dummy line like `foo = 1`)
  - commit the change
    ```bash
    git add tests/models/bert/test_modeling_bert.py
    git commit -m "dummy commit"
     ```
  - run the test fetcher against the previous commit
    ```bash
    python utils/tests_fetcher.py  --diff_with_last_commit
    ```
- Now, the results:
  TL;DR: `test_modeling_bert.py` is not included by the new version of test fetcher. But I think it should be included.

  - on `main`
  (`tests/models/bert/test_modeling_bert.py` is in `TEST TO RUN` and in the file `test_list.txt`)
    ```
    ### DIFF ###

    ### MODIFIED FILES ###
    - tests/models/bert/test_modeling_bert.py

    ### IMPACTED FILES ###
    - tests/models/auto/test_modeling_auto.py
    - tests/models/auto/test_modeling_tf_auto.py
    - tests/models/bert/test_modeling_bert.py
    - tests/models/encoder_decoder/test_modeling_encoder_decoder.py
    - tests/models/speech_encoder_decoder/test_modeling_speech_encoder_decoder.py
    - tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py
    - tests/models/vision_text_dual_encoder/test_modeling_vision_text_dual_encoder.py

    ### TEST TO RUN ###
    - tests/models/auto/test_modeling_auto.py
    - tests/models/auto/test_modeling_tf_auto.py
    - tests/models/bert/test_modeling_bert.py
    - tests/models/encoder_decoder/test_modeling_encoder_decoder.py
    - tests/models/speech_encoder_decoder/test_modeling_speech_encoder_decoder.py
    - tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py
    - tests/models/vision_text_dual_encoder/test_modeling_vision_text_dual_encoder.py
    ```

  - on `test_fetch_v2`
    (`tests/models/bert/test_modeling_bert.py` is **NEITHER** in `TEST TO RUN`, **NOR** in the file `test_list.txt`)
    ```
    ### MODIFIED FILES ###
    - tests/models/bert/test_modeling_bert.py

    ### IMPACTED FILES ###
    - tests/models/auto/test_modeling_auto.py
    - tests/models/auto/test_modeling_tf_auto.py
    - tests/models/bert/test_modeling_bert.py
    - tests/models/encoder_decoder/test_modeling_encoder_decoder.py
    - tests/models/speech_encoder_decoder/test_modeling_speech_encoder_decoder.py
    - tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py
    - tests/models/vision_text_dual_encoder/test_modeling_vision_text_dual_encoder.py

    ### TEST TO RUN ###
    - tests/models/auto/test_modeling_auto.py
    - tests/models/auto/test_modeling_tf_auto.py
    - tests/models/encoder_decoder/test_modeling_encoder_decoder.py
    - tests/models/speech_encoder_decoder/test_modeling_speech_encoder_decoder.py
    - tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py
    - tests/models/vision_text_dual_encoder/test_modeling_vision_text_dual_encoder.py
    ```

**Comment by ydshieh** [comment]
Other than the above question, could you also show me (us) how to see the following in action (on the `main` branch)🙏 

> As a comparison, the previous version stopped at transformers/__init__.py.

I think this happens internally, not on the final outputs. I would like to know at which places this happens - the goal is to understand things better if you have sometime to share.

Thanks.

**Comment by ydshieh** [comment]
Well,  at least, when `src/transformers/models/bert/modeling_bert.py` is changed, the test file `tests/models/bert/test_modeling_bert.py` included 👍 . So the dependency detection seems to work well, and the above situation is just an edge case (to including self)

**Review by amyeroberts — APPROVED** [review]
No comments other than to say this was a very interesting and enjoyable PR to review ❤️  

Thanks for adding!

File: tests/repo_utils/test_tests_fetcher.py [hunk 3]
```diff
@@ -13,52 +13,654 @@
...
+                "tests/test_modeling_common.py",
+                "tests/models/bert/test_modeling_bert.py",
+                "tests/models/gpt2/test_modeling_gpt2.py",
+            }
+            assert set(test_map["src/transformers/__init__.py"]) == expected_init_tests
+
+    def test_infer_tests_to_run(self):
+        with tempfile.TemporaryDirectory() as tmp_folder:
+            tmp_folder = Path(tmp_folder)
+            models = models = ["bert", "gpt2"] + [f"bert{i}" for i in range(10)]
```
[the remark this answers is not in the document]
Comment (sgugger) [reply tests/repo_utils/test_tests_fetcher.py:573]: Just a typo ;-)

**Comment by sgugger** [comment]
@ydshieh, good catch on a modified test file missing from the tests launched. I have only put the dependencies and forgot those. Will fix.

File: utils/tests_fetcher.py [hunk 4]
```diff
@@ -24,13 +45,36 @@
...
-all_model_test_files = [str(x) for x in Path("tests/models/").glob("**/**/test_modeling_*.py")]
-
-all_pipeline_test_files = [str(x) for x in Path("tests/pipelines/").glob("**/test_pipelines_*.py")]
+PATH_TO_REPO = Path(__file__).parent.parent.resolve()
+PATH_TO_TRANFORMERS = PATH_TO_REPO / "src/transformers"
+PATH_TO_TESTS = PATH_TO_REPO / "tests"
+
+# List here the models to always test.
+IMPORTANT_MODELS = [
+    # Most downloaded models
```
[the remark this answers is not in the document]
Comment (sgugger) [reply utils/tests_fetcher.py:54]: Not sure people are going to go read this periodically ;-) It's more of us to think of it when we add a new pipeline for instance.

**Comment by sgugger** [comment]
@ydshieh did you want to review more or is it good to merge?

**Comment by ydshieh** [comment]
> @ydshieh did you want to review more or is it good to merge?

Hi @sgugger If you feel urgent to merge, go ahead. Otherwise, I would love to continue the review process despite I am slow.

File: utils/tests_fetcher.py [hunk 5]
```diff
@@ -166,96 +190,173 @@ def get_diff(repo, base_commit, commits):
...
+
+def extract_imports(module_fname, cache=None):
     """
-    Get the dependencies of a module.
+    Get the imports a given module makes. This takes a module filename and returns the list of module filenames
+    imported in the module with the objects imported in that module filename.
     """
-    with open(os.path.join(PATH_TO_TRANFORMERS, module_fname), "r", encoding="utf-8") as f:
+    if cache is not None and module_fname in cache:
+        return cache[module_fname]
```
Comment (ydshieh) [inline utils/tests_fetcher.py:245]: Very nice! I actually missed these 2 lines during the review, and I was thinking why not using cache but just adding content at the end.

File: utils/tests_fetcher.py [hunk 6]
```diff
@@ -308,290 +413,159 @@ def create_reverse_dependency_map():
...
-                    else:
-                        test_files_to_run.extend(new_tests)
-
-        # Remove duplicates
+        # All modified tests need to be run.
+        test_files_to_run = [
+            f for f in modified_files if f.startswith("tests") and f.split(os.path.sep)[-1].startswith("test")
+        ]
+        # Then we grab the corresponding test files.
+        test_map = create_module_to_test_map(reverse_map=reverse_map, filter_models=filter_models)
```
Comment (ydshieh) [inline utils/tests_fetcher.py:553]: good not to create `reverse_map` twice!

File: utils/tests_fetcher.py [hunk 7]
```diff
@@ -704,33 +669,54 @@ def filter_tests(output_file, filters):
...
+        repo = Repo(PATH_TO_REPO)
+        commit_message = repo.head.commit.message
+        commit_flags = parse_commit_message(commit_message)
+        if commit_flags["skip"]:
+            print("Force-skipping the CI")
+            quit()
+        if commit_flags["test_all_models"]:
+            print("Testing all models found.")
+        if commit_flags["test_all"]:
+            print("Force- launching all tests")
```
Comment (ydshieh) [inline utils/tests_fetcher.py:693]: When reading this part, I feel confused between between `"test_all_models"` and `"test_all"`.
Maybe `run_fetched_tests` and `run_all_tests` make it more clear. 
Leave you to decide.

(and `"Run all fetched tests.`)

File: tests/repo_utils/test_tests_fetcher.py
Comment (ydshieh) [inline tests/repo_utils/test_tests_fetcher.py:1]: 😮 💯 🔥

**Review by ydshieh — COMMENTED** [review]
Thank you @sgugger  again for the great work and the patience for my slow review.

I am actually learning things here rather than giving useful reviews, but left 2 nits (typo and var. naming).

**Review by ydshieh — APPROVED** [review]

File: utils/tests_fetcher.py (hunk 7 above)
Comment (sgugger) [reply utils/tests_fetcher.py:693]: Going for no_filter instead, as I find it clearer and shorter than `run_fetched_tests`.
